### ISSUE TITLE
Libai-chat

### ISSUE THREAD

**Issue by ChenQiaoling00** [ChenQiaoling00/aliyun_ml#1]
## 1. Basic

libai-chat 是融合了 deepspeed chat 和 colossalai  chat 的 RLHF 框架的一个开源项目。

libai-chat 回报模型的结构和 loss 设计基本和 Deepspeed chat 保持一致，分数是取句末 token。的分数。然而 ColossalAI的回报模型分数是将句子所有的token求平均。
    即 compute_rewards 和 get_advantages_and_returns 与 Deepspeed chat 保持一致。
libai-chat 训练流程和 colossalAI 保持一致，即 _make_experience 和 train step（历史遗留问题，一开始搭架构的时候是参考 colossalAI）。

影响训练的超参数,在 config/opt_ppo.py 里定义：
```bash
        ppo_kl_coef=1e-4,
        ppo_gamma=1,
        ppo_lam=0.95,
        ppo_cliprange=0.2,
        ppo_value_cliprange=0.2,
        ppo_vf_coef=0.1,
        num_episodes=10000,
        max_steps=10,
        update_steps=10,
```

## 问题
1. datasets 的问题

2. OneFlow 特有问题

3. Reward model 的问题

4. 超参数的问题，比如学习率。

### 1. datasets:
|  表头   | deepspeed  |  libai-chat  | libai-chat  | libai-chat  |libai-chat  |
|  ----  | ----  | ----  | ----  | ----  |----  |
| actor dataset  | Dahoas/rm-static | instinwild_en_train | instinwild_en_train | instinwild_en_train |instinwild_en_train |
| rw dataset  | Dahoas/rm-static |hh-rlhf| hh-rlhf| Dahoas/rm-static|instinwild_en_train |
|ppo dataset| Dahoas/rm-static| hh-rlhf |Dahoas/rm-static| Dahoas/rm-static|instinwild_en_train |
|reward score|increase at 800 steps|stable utils 1000 step|stable utils 1000 step|stable utils 1000 step| loss 激增 |

1.1 其中libai -chat 的set 1 和 set2 都是在 rw 和 ppo 阶段用的是同样的数据集，分别是 hh-rlhf 和 Dahoas/rm-static，此时 reward score 最高点到 0.1 并大多数是负数。
<img width="650" alt="image" src="https://github.com/ChenQiaoling00/aliyun_ml/assets/123612019/f30eca99-3a78-4f7d-b844-1e1f2906dcc8">

1.2 libai-chat 的 set 3 的 rw 和 ppo 和 sft 阶段用的都是不同的数据集，分别是 hh-rlhf和 Dahoas/rm-static 和 instinwild_en_train，此时 reward score 都为正数。
<img width="557" alt="image" src="https://github.com/ChenQiaoling00/aliyun_ml/assets/123612019/f3860cdc-5d51-4ada-891f-c2debd3de7d9">

1.3 libai-chat 的 set 4 的 rw 和 ppo 和 sft 阶段用的都是相同的数据集为 instinwild_en_train，此时 loss 的值 在 1000 个step 后激增，reward score 在 上升后下降。
![image](https://github.com/ChenQiaoling00/aliyun_ml/assets/123612019/2c155d51-9174-435a-b60f-c170e491b979)

1.4 Deepspeed chat 的 rw 和 ppo 阶段用的都是相同的数据集为 Dahoas/rm-static 的不同部分，此时 reward score 在 800 个step 后稳定上升。（无论实验多少次都是这样）
![image](https://github.com/ChenQiaoling00/aliyun_ml/assets/123612019/b7f9b52b-e01a-41d1-959f-424ede619898)

**建议重写一下 dataloader 使之可以讲一个数据集按比例分成三份，分别给 step1-sft，step2-RM，step3-ppo 使用。**

### 2. OneFlow 特有问题
2.1 在做数值正确性验证时，在计算 reward 时发现，oneflow 这样做不会改变 tensor 的值，torch 这样是可以的.
原因是 global tensor 的 indexing 操作没有返回 view。所以需要绕过。
```bash
>>> placement=flow.placement(type="cuda", ranks=[0])
>>> sbp=(flow.sbp.broadcast,)
>>> rewards=flow.ones([4,512],placement=placement,sbp=sbp)
>>> reward_clip=flow.ones([4],placement=placement,sbp=sbp)
>>> rewards[0,311:511][-1]+=reward_clip[0]
>>> rewards[0,311:511][-1]
tensor(1., placement=oneflow.placement(type="cuda", ranks=[0]), sbp=(oneflow.sbp.broadcast,), dtype=oneflow.float32)
```

2.2 OneFlow 暂时没有显存优化的方法
所以在单机单卡 data parallel = 1 训练时， batch size 设为 4，显存占用约为 70G，相比于 Deepspeed chat 的单机仅占用 20G，显存占用增加了 50G。
故开发机器只能在 A100 80G 上训练，而不能在 3090 24G 上训练。

### 3. Reward model 的问题
由于 reward score 和 advatage 一开始很稳定的在 0 附近浮动，所以认为可能是 libai-chat 在训练 reward model 时没有收敛，导致 reward model 的能力不行。

故训练 reward model 到 loss 收敛，用最新的收敛的 reward model 再次重训 ppo，结果没有改善。

（尽管 original reward model 是直接 download huggingface 的 facebook/opt-350m 预训练模型，照理来说 reward model 的能力应该是足够的，并且 deepspeed chat的 reward model 仅训练 3 个 step 就能使 ppo 阶段收敛）。

### 4. 超参数的问题，比如学习率。
4.1 学习率

colossalai 的 学习率为 1e-7，尝试使用 deepspeed 的学习率 optim.lr = 9.65e-6 重训练，结果 reward score 的结果仍然在 0 附近浮动。
